### PR TITLE
[WIP] Implement Search

### PR DESCRIPTION
**Pull request by aero31aero**
**This is just a work in progress PR to let the community know this is being worked on.**

Implements search and starts migration to promises. Adds the following new dependencies;

1. Bluebird: refer #160
2. Natural: for tokenization, stemming and tfidf calculation.
3. Glob: for easy selection of all cached pages. This could be removed at a later stage with a custom implementation, but for dev, I'll let it stay.

Please review this checklist before submitting a pull request.

- [x] Code compiles correctly
- [x] Created tests, if possible
- [x] All tests passing (`npm run test:all`)
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

**Comment by agnivade** [comment]
Thanks for the PR. Please have a look into my [comment](https://github.com/tldr-pages/tldr-node-client/issues/160#issuecomment-343799438)

About the search functionality, is it like the `apropos` command ? i.e. Does it search for page contents also and return the command names whose pages contain those keywords ?

**Comment by aero31aero** [comment]
Yes, like the `apropos` command. Would search within the manual pages and implement basic information retrieval techniques to try to fetch relevant results.

**Comment by agnivade** [comment]
Ok, please use native promises instead of bluebird. And use `let` instead of `var`. 

I will do a review once you add tests and documentation. Thanks.

**Comment by aero31aero** [comment]
Initial implementation now looks complete. Here are the usage instructions:

1. `tldr update`. This would also create the corpus.
2. `tldr search "create symbolic link to file"`.

You need to put the query in quotes if the query contains any defined option as well. For example, `npm start -- search file search` would fail. Looking for a way around that in the current options library.

Still not ready for a review but I'd like to ask what kind of UI would be better. As of now, I just `console.log` the top 10 results as filenames of the cached files and their rank. I had three ideas:

1. Print the top page.
2. List the top 10 results such that typing `tldr <result>` would then get the content of the page.
3. Print the top page and then list the top 10 results.

Feedback appreciated.

**Comment by agnivade** [comment]
Please go with option 2.

Also, how about using an ES6 class for this ? It seems to be particularly well suited for this purpose.

File: package.json [hunk 1]
```diff
@@ -41,7 +41,7 @@
     "example": "node ./bin/tldr tar",
     "test": "mocha test --require=env-test",
     "test:quiet": "npm test --reporter=dot",
-    "lint": "eslint lib test bin/tldr",
+    "lint": "eslint --global Promise lib test bin/tldr",
```
Comment (agnivade) [inline package.json]: Please set this in the `.eslintrc`. Avoid putting settings from the command line.

Comment (agnivade) [reply package.json]: Taken care of in #166. You can remove this.

**Comment by aero31aero** [comment]
Explain what do you mean by using a class here. You mean convert the whole module to a class? I thought that since this file is just one module exporting only two endpoints, it should be fine.

And :+1: for option 2.

File: package.json (hunk 1 above)
Comment (aero31aero) [reply package.json]: Apologies for this. Didn't notice the `.eslintrc` file earlier.

**Comment by agnivade** [comment]
>  You mean convert the whole module to a class?

Yes. But now that I think of it, you would have to create the object for this in tldr.js and then call the functions. And the object will just remain in global scope. Not really elegant. We can drop the idea. I will revisit it later when we are done with full promisification of the codebase.

**Comment by aero31aero** [comment]
Ready for final review now. I intend to make small incremental improvements overtime. Also, since most of the functions that can be tested are private, I'd have to come up with a testing strategy for them. Perhaps a separate PR for it if this gets merged.

Sample output:

```bash
$ tldr --search create symbolic link to file --os windows
Searching for: create symbolic link to file

    $ ln
    $ update-alternatives (Available on: linux)
    $ nm (Available on: linux, osx)
    $ gcc
    $ chgrp
    $ c99
    $ apg
    $ mkdir
    $ umask (Available on: linux)

Run tldr <command> to see specific pages.
```

**Comment by agnivade** [comment]
We need to have some basic sanity tests atleast. No need to rush through anything. Take your time, come  up with a proper testing strategy. And then we can get to reviewing this.

Thanks again for all your work.

**Comment by agnivade** [comment]
Btw, I have promisified `index.getShortIndex`. So you will need to update your code slightly.

**Comment by aero31aero** [comment]
I was just worried the PR was getting too big and thought would add the tests in another one.

Anyway, pushed the basic integration tests. Have exams soon so would not be able to work on the other tests for some time. If you want to merge only after all the tests are there, I'll write them sometime in December. That was kind of why I was hoping to get this done quick. :P

Also, rebased with latest master and updated for latest changes.

**Comment by agnivade** [comment]
Great. I will take it from here. You can add more tests later.

**Comment by agnivade** [comment]
Closing in favor of #171.

**Pull request closed without merging**